## Computational Linguistics group assignment

In this assignment you will be asked to extend the work by Gatti et al by checking whether form-meaning mappings learned on a different yet related language to that considered in the original study still capture the perceived valence of pseudowords. To do this you will be asked to engage with several different resources and adapt the pipeline following the instructions. Along the way, you will be asked to answer a few questions.

You need to submit the complete notebook in .ipynb format, with intermediate outputs visible. The notebook should be named as follows:

CL2025_groupN_assignment.ipynb

where N is the group number. Submissions in the wrong format or with names not adhering to the guidelines will not be evaluated.

Indicate group members' names, student numbers, and contributions below:
- 1. Barbora Hájková (2093396) - Worked on task 1 and 4
- 2. Frey Mink (2105524) - Worked on task 1, 5 and 6
- 3. Gaia Naumann (2091218) - Worked on task 2, 7 and 8
- 4. Stefan Hasenoot (2091624) - Worked on task 3, 7 and 8


Comments made by any groupmembers are recognizable by the "##" in front. Singular "#" indicates a comment left as an instruction.

In [ ]:
## This block has been modified to work for local usage 

# the code has been tested using the psycho-embeddings library to extract representations from LLMs. You can also use other libraries,
# as long as you make sure that you are producing the correct output.


#git clone https://github.com/MilaNLProc/psycho-embeddings.git
%pip install datasets
#pip install fasttext
%pip install nltk
%cd psycho-embeddings

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Note: you may need to restart the kernel to use updated packages.


huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


Note: you may need to restart the kernel to use updated packages.
/home/gaianaumann/Documents/GitHub/CL_group8/psycho-embeddings


In [465]:
# the solution to the assignment has been obtained using these packages.
# you're free to use other packages though: consider this as an indication, not a prescription.
import nltk
import numpy as np
import pandas as pd
import fasttext as ft
import pickle as pkl
import fasttext.util
from tqdm import tqdm
from collections import defaultdict
from transformers import AutoTokenizer
#from psycho_embeddings import ContextualizedEmbedder


**Task 1** (*10 points available, see breakdown per task below*)

You should replicate the main design in the paper *Valence without meaning* by Gatti and colleagues (2024), using estimates collected for Dutch word valence to train linear regression models and apply them to predict the valence of English pseudowords from Gatti and colleagues.

In detail, to train your regression models, you should use the dataset by Speed and Brysbaert (2024) containing crowd-sourced valence ratings (use the metadata to identify the relevant columns) collected for approximately 24,000 Dutch words. See the paper *Ratings of valence, arousal, happiness, anger, fear, sadness, disgust, and surprise for 24,000 Dutch words* by Speed and Brysbaert (2024).

You should train a letter unigram model and a bigram model. Each model should be trained on Dutch words only.

Pay attention to one issue though: pseudowords created for English may be valid words in Dutch: therefore, you should first filter the list of pseudowords against a large store of Dutch words. To do so, use the words in the Dutch prevalence lexicon available in this OSF repository: https://osf.io/9zymw/. Essentially, you need to exclude any pseudoword that happens to be a word for which a prevalence estimate is available, whatever the prevalence is.

Each code block indicates how many points are available and how they are attributed.

In [380]:
# read in the pseudowords from Gatti and colleagues, as well as the valence ratings for 24,000 Dutch words from Speed and Brysbaert (2024)
# show the first 5 lines of each dataset.
# 1 point for identifying the correct files and correctly loading their content

In [ ]:
## getting the pseudowords from Gatti

df_Gatti = pd.read_csv("Gatti_pseudowords_data_2_full.csv")
df_Gatti.head(5)

,Unnamed: 0,X,pseudoword,Value,predicted_valence,predictedL_valence,predictedL_Bi_valence,predicted_Dim_valence,predictedL_Dim_valence,predictedBi_Dim_valence,predictedBi_valence,LDist,Ortho_VAL,Semant_Neigh,SDist,Semant_VAL
0,0,1,abhert,0.452501,7.414814,5.116167,6.444633,6.783771,6.630497,7.414814,6.444633,2,4.655714,ordinary,0.558492,5.05
1,1,2,abhict,0.434171,8.233714,5.059183,6.509936,7.366068,7.377534,8.233714,6.509936,2,3.093333,cardigan,0.622202,5.95
2,2,3,acleat,0.527803,5.552468,5.262971,5.245826,5.268643,5.396114,5.552468,5.245826,1,4.240000,solarium,0.575150,6.10
3,3,4,acmure,0.604889,8.714640,5.120029,6.562896,7.680827,7.583230,7.809910,5.414532,2,5.885000,bad,0.570299,3.24
4,4,5,acoed,0.538990,7.340002,5.115652,5.309727,7.105662,7.024771,7.340002,5.309727,1,5.680000,girl,0.499035,7.15


In [46]:
## Getting the word valence from Speed

df_Speed = pd.read_csv("wordvalence_Speed.csv")
df_Speed.head(5)

,Word,Arousal,Valence,ValenceCategory,ValenceVsNeutral,Happiness,Anger,Fear,Sadness,Disgust,...,Length,Nsyl,N_phonemes,PoS,OLD20,AoA,DLP_RT,DLP_Acc,DCP_RT,DCP_Acc
0,mama,2.812500,4.000000,positive,valenced,3.300000,1.000000,1.000000,1.100000,1.000000,...,4,2,4,N,1.55,2.044257,513.530256,1.000000,931.03,0.99
1,ja,2.823529,3.894737,positive,valenced,3.818182,1.090909,1.181818,1.181818,1.000000,...,2,1,2,Function,1.00,2.250000,494.980294,1.000000,1014.77,1.00
2,papa,2.562500,3.722222,positive,valenced,4.142857,1.142857,1.000000,1.000000,1.000000,...,4,2,4,N,1.65,2.336327,570.497647,1.000000,936.88,0.99
3,nee,2.928571,2.350000,negative,neutral,1.000000,1.727273,1.363636,1.454545,1.363636,...,3,1,2,Function,1.00,2.555556,488.637879,0.902439,1088.01,1.00
4,kaka,3.357143,2.050000,negative,neutral,1.090909,1.454545,1.181818,1.000000,4.727273,...,4,2,4,N,1.80,2.611111,629.402647,0.850000,1605.49,0.53


In [384]:
# filter out pseudowords that happen to be valid Dutch words (mind case folding!)
# show the set of pseudowords filtered out.
# 1 point for applying the correct filtering

In [ ]:
## Combining the file for the prevalence of Dutch words and Belgian words
df_prevalence_Netherlands = pd.read_csv("prevalence_netherlands.csv" , sep = "\t")
df_prevalence_Belgium = pd.read_csv("prevalence_belgium.csv", sep= "\t")

df_prevalence_combined = pd.concat([df_prevalence_Belgium, df_prevalence_Netherlands], join= "outer") ##outer: takes union

## Checking if they all have a prevalence value
df_prevalence_combined_filtered = df_prevalence_combined[df_prevalence_combined["prevalence"] != None]

## Saving only the column with words as a Series
df_Dutch_words = df_prevalence_combined_filtered.word
df_Dutch_words = df_Dutch_words.drop_duplicates()

## Applying filtering and showing the set of filtered out pseudowords.
print(df_Gatti.pseudoword[df_Gatti.pseudoword.isin(df_Dutch_words)])
df_Gatti_clean = df_Gatti[~df_Gatti.pseudoword.isin(df_Dutch_words)]

## Saving only the column with pseudowords from pseudowords_Gatti as a Series
df_pseudowords = df_Gatti_clean.pseudoword

900    pimpen
Name: pseudoword, dtype: object


In [388]:
# encode Dutch words and pseudowords from Gatti et al as uni- and bi-gram vectors
# show the uni-gram and bi-gram encoding of the pseudoword ampgrair
# 2 points for correctly encoding the target strings as uni- and bi-gram vectors

In [48]:
## this is simplified code from the notebook of class one. Permission has been given to use it with the correct reference.

def ngram_featurizer(s, n):

    string_boundary = ["letter"]*(n-1)                        ## necessary to encode features such as 'this string begins/ends with this specific symbol'
    s = string_boundary + list(s) + string_boundary

    return [tuple(s[i:i+n]) for i in range(len(s)-n+1)]       ## this is where the n-gram featurization actually happens.


In [52]:
## this is code adapted from the notebook of class one. Permission has been given to use it with the correct reference.

## This function encodes all strings as frequency counts over n-grams
def encode_corpus(corpus, n, mapping=None, dictionary=False):

    ## creates the appropriate mapping dictionary if it does not exist yet
    if not mapping:
        all_ngrams = set()
        for instance in corpus:
            ## get a comprehensive set of all n-grams in the corpus
            all_ngrams = all_ngrams.union(
                set(ngram_featurizer(instance, n))
                )

        ## map each n-gram to an integer which will index the feature matrix
        mapping = {ngram: i for i, ngram in enumerate(sorted(all_ngrams))}


    if dictionary == True:
    ## creates a dictionary with as key the pseudoword and as value the ngrams present

        dictionary = {}
        for pseudoword in df_pseudowords:
            lst = [0 for i in range(len(mapping))]

            ngrams = ngram_featurizer(pseudoword, n)
            for ngram in ngrams:
                ngram_index = mapping[ngram]
                lst[ngram_index] += 1

            dictionary[pseudoword] = lst    

        return dictionary, mapping
    

    if dictionary == False:
        ## create a feature matrix of the appropriate dimensionality

        word_list = np.zeros((len(corpus), len(mapping)))
        for i, instance in enumerate(corpus):
            for ngram in ngram_featurizer(instance, n):
                try:
                    ## access the right column given the n-gram being processed
                    word_list[i, mapping[ngram]] += 1
                except KeyError:
                    ## if the current n-gram is new, skip it
                    pass

    return word_list, mapping



dictionary_unigram, mapping_unigram = encode_corpus(list(df_pseudowords), 1, dictionary=True)
dictionary_bigram, mapping_bigram = encode_corpus(list(df_pseudowords), 2, dictionary=True)

In [ ]:
## showing the uni-gram and bi-gram encoding of the pseudoword ampgrair
print(dictionary_unigram["ampgrair"])
print(dictionary_bigram["ampgrair"])

[2, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [57]:
##getting words from df_speed

speed_words = df_Speed["Word"]
print(speed_words.head(5))

speed_valence = df_Speed["Valence"] / 5 ## Divided by 5 to normalize the scale (1 through 5)
print(speed_valence.head(5))

0    mama
1      ja
2    papa
3     nee
4    kaka
Name: Word, dtype: object
0    0.800000
1    0.778947
2    0.744444
3    0.470000
4    0.410000
Name: Valence, dtype: float64


In [ ]:
## Encoding speed_words into unigrams and bigrams
speed_unigram, word_mapping_unigram = encode_corpus(list(speed_words), 1, mapping_unigram)
speed_bigram, word_mapping_bigram = encode_corpus(list(speed_words), 2, mapping_bigram)

In [ ]:
## encoding pseudowords as unigrams and bigrams
## uses the same mapping as word uni/bi grams
pseudoword_unigrams, pseudoword_mapping_unigram = encode_corpus(list(df_pseudowords), 1, mapping = word_mapping_unigram)
pseudoword_bigrams, pseudoword_mapping_bigram = encode_corpus(list(df_pseudowords), 2, mapping = word_mapping_bigram)

In [396]:
# use word valence estimates from Speed and Brysbaert (2024) to train
# - a uni-gram model
# - a bi-gram model
# 2 points for correctly trained models

In [56]:
## fit a linear model (fits speed_valence to speed_unigrams) 
from sklearn.linear_model import LinearRegression

lm_unigrams = LinearRegression()
lm_unigrams.fit(speed_unigram, speed_valence)

LinearRegression()

In [398]:
## fit a linear model (fits speed_valence to speed_bigrams) 
from sklearn.linear_model import LinearRegression

lm_bigrams = LinearRegression()
lm_bigrams.fit(speed_bigram, speed_valence)

LinearRegression()

In [399]:
# apply trained models to predict the valence of pseudowords from Gatti et al (2024).
# Then apply the same models back onto the training set to see how well they predict the valence of words in Speed and Brysbaert (2024).
# 2 points for correctly applied models

In [ ]:
## uses LM_unigrams to predict valence of pseudowords
pw_unigram_valence_pred = lm_unigrams.predict(pseudoword_unigrams)
pw_unigram_valence_pred

array([0.58388796, 0.58702303, 0.60076878, ..., 0.56825794, 0.56783317,
       0.55887553], shape=(1499,))

In [ ]:
## uses LM_bigrams to predict valence of pseudowords
pw_bigram_valence_pred = lm_bigrams.predict(pseudoword_bigrams)
pw_bigram_valence_pred

array([0.66424937, 0.62474322, 0.6470547 , ..., 0.57171145, 0.5836004 ,
       0.6096178 ], shape=(1499,))

In [ ]:
# Applies the same models back onto the training set to see how well they predict the valence of words in Speed and Brysbaert (2024).
word_unigram_valence_pred = lm_unigrams.predict(speed_unigram)
print(word_unigram_valence_pred)
word_bigram_valence_pred = lm_bigrams.predict(speed_bigram)
print(word_bigram_valence_pred)

[0.59484292 0.60994254 0.59963784 ... 0.59224499 0.59018619 0.58212288]
[0.63318787 0.60696509 0.64496898 ... 0.54409183 0.53076193 0.60158886]


In [403]:
# compute the Spearman correlation coefficients between true valence and predicted valence under both uni- and bi-gram models for
# - words from Speed and Brysbaert (2024)
# - pseudowords from Gatti and colleagues (2024)
# show both correlation coefficients.
# 2 points for the correct Spearman correlation coefficients (rounded to the third decimal place)

In [404]:
## computing spearman's r between predicted word valence and true word valence
from scipy.stats import spearmanr
print(spearmanr(word_unigram_valence_pred, speed_valence))
print(spearmanr(word_bigram_valence_pred, speed_valence))

SignificanceResult(statistic=np.float64(0.08747282146010639), pvalue=np.float64(5.82250941925177e-42))
SignificanceResult(statistic=np.float64(0.3125232744970933), pvalue=np.float64(0.0))


In [405]:
## computing spearman's r between predicted pseudword valence and true pseudoword valence
from scipy.stats import spearmanr
true_pseudoword_valence = df_Gatti_clean["predicted_valence"] / 9 # Divided by 9 to normalize the 9-point scale
print(spearmanr(pw_unigram_valence_pred, true_pseudoword_valence))
print(spearmanr(pw_bigram_valence_pred, true_pseudoword_valence))

SignificanceResult(statistic=np.float64(0.32970178214486584), pvalue=np.float64(2.4293751061479433e-39))
SignificanceResult(statistic=np.float64(0.33998397507550654), pvalue=np.float64(7.098906611226265e-42))


**Task 2** (*8 points available, see breakdown below*)

Again following Gatti and colleagues, you should encode the target strings (pseudowords from Gatti and Dutch words from Speed and Brysbaert) as fastText embeddings, train a multiple regression model on Dutch words and apply it to the pseudowords in Gatti et al. You should finally report the Spearman correlation coefficient between observed and predicted valence for both words and pseudowords.

You should use the pre-trained fastText model for Dutch, available at this page: https://fasttext.cc/docs/en/crawl-vectors.html

Finally, you should answer two questions about the fastText model (see below).

In [ ]:
# load the fastText model
# 1 point for correctly loading the appropriate fastText model

##downloaded model from the internet due to long waiting times
#ft_nl = fasttext.load_model('/home/gaianaumann/Documents/cc.nl.300.bin')

##The code below would be the correct way to download the model through code. 
import fasttext.util
fasttext.util.download_model('nl', if_exists='ignore')  
ft_nl = fasttext.load_model('cc.nl.300.bin')



What is the dimensionality of the pre-trained Dutch fastText embeddings? (*1 point for the correct answer*)

> The pretrained FastText word vector embeddings have a dimensionality of 300. 

What minimum and maximum n-gram size was specified for training this fastText model? (*1 point for the correct answer*)

> The minimum n-gram size specified is 3 and the maximum is 6. These are the default values and could be overwritten.

In [ ]:
# encode Dutch words and pseudowords as fastText embeddings
# show the first 20 values of the embedding of the word 'speelplaats' and of the pseudoword 'danchunk'
# 2 points for correctly encoding words and pseudowords with fastText

##Creates embeddings for the Dutch words
Speed_Emb_List = [ft_nl.get_word_vector(i) for i in df_Speed['Word']]

##Creates embeddings for the pseudowords
Pseudo_Emb_List = [ft_nl.get_word_vector(i) for i in df_pseudowords] 

##Makes sure the words are tied to their corresponding embeddings
df_words_FastText = pd.DataFrame(data={"word":df_Speed['Word'],"embedded":Speed_Emb_List})
df_pseudowords_FastText = pd.DataFrame(data={"word":df_pseudowords,"embedded":Pseudo_Emb_List})

##Takes the first 20 embeddings for both the normal Dutch word and the pseudoword and prints them
print("speelplaats:", end=" ")
print([emb for emb in df_words_FastText[df_words_FastText.word=="speelplaats"].embedded][0][:20])

print("danchunk:", end=" ")
print([emb for emb in df_pseudowords_FastText[df_pseudowords_FastText.word=="danchunk"].embedded][0][:20])


speelplaats: [ 0.0253247  -0.00634261  0.02746305 -0.04024595  0.04888906  0.00660965
 -0.04152017 -0.01824508 -0.00645641  0.00093806  0.0708492  -0.03291791
  0.00263817 -0.02825846 -0.02188046 -0.03188037 -0.01846142 -0.02203094
 -0.01883078 -0.00259199]
danchunk: [-0.00592199  0.00097547  0.05925412  0.00053251 -0.00386978 -0.02089076
 -0.02829577  0.00972911 -0.02510111 -0.11454885 -0.02695064  0.01551034
  0.02384409  0.01009528  0.04545438  0.00997385 -0.00474529  0.02524533
  0.02430548 -0.02851078]


In [408]:
# train regression model on word valence
# 1 point for correctly training the regression model

Dutch_valence_FastText = LinearRegression()
Dutch_valence_FastText.fit([i for i in df_words_FastText['embedded']], speed_valence)


LinearRegression()

In [409]:
# apply the trained model to predict the valence of pseudowords from Gatti et al (2024).
# Then apply the same model back onto the training set to see how well it predicts the valence of words in Speed and Brysbaert (2024).
# 1 point for correctly applied model

pseudo_valence_FastText_pred = Dutch_valence_FastText.predict([i for i in df_pseudowords_FastText['embedded']])
print(pseudo_valence_FastText_pred)

Dutch_valence_FastText_pred = Dutch_valence_FastText.predict([i for i in df_words_FastText['embedded']])
print(Dutch_valence_FastText_pred)

[0.6231049  0.6153373  0.5770528  ... 0.6034669  0.58416224 0.58479464]
[0.8619921  0.57310563 0.8073086  ... 0.63601154 0.5615791  0.62461823]


In [410]:
# compute the Spearman correlation coefficients between true valence and predicted valence for
# - words from Speed and Brysbaert (2024)
# - pseudowords from Gatti and colleagues (2024)
# show the correlation coefficient.
# 1 point for the correct Spearman correlation coefficients (rounded to the third decimal place)

In [ ]:
##performs spearman on test set data predictions (Gatti)
print(spearmanr(pseudo_valence_FastText_pred, true_pseudoword_valence))

##performs spearman on training set data predictions (Speed)
print(spearmanr(Dutch_valence_FastText_pred, speed_valence))

SignificanceResult(statistic=np.float64(0.17641512677343416), pvalue=np.float64(6.045675204139514e-12))
SignificanceResult(statistic=np.float64(0.7243392863663819), pvalue=np.float64(0.0))


**Task 3** (*6 points available, see breakdown below*)

Now you are asked to extend the work by Gatti et al by also considering the representations learned by a transformer-based models, in detail *RobBERT v2* (https://huggingface.co/pdelobelle/robbert-v2-dutch-base). You should follow the same pipeline as for the previous models, encoding both Dutch words from Speed and Brysbaert (2024) and the pseudowords from Gatti et al using the embedding of each string at layer 0, before positional information is factored in. If a string consists of multiple tokens, average the embeddings of all tokens to produce the embedding of the whole string. Then train a multiple regression model on the valence of Dutch words, apply it to the pseudowords, and compute the Spearman correlation between observed and predicted ratings.

Use the HuggingFace model card for RobBERT v2 to check how to access it.

I recommend saving the embeddings to file once you have generated them and you know they are correct: embedding thousands of strings takes some time, and you don't want to have to do it again. For the same reason, develop your code by considering only a small fractions of the words and pseudowords, in order to quickly see if something is wrong. Only when you are positive it works, embed all strings.

In [412]:
# load and instantiate the right model
# 1 point for loading the right model
from transformers import AutoTokenizer, AutoModelForMaskedLM

tokenizer = AutoTokenizer.from_pretrained("pdelobelle/robbert-v2-dutch-base")
model = AutoModelForMaskedLM.from_pretrained("pdelobelle/robbert-v2-dutch-base")

In [413]:
# encode the words and pseudowords using RobBERT v2. I've used the free GPU runtime on COLAB to speed things up,
# but in this case you need to batch the words and pseudowords. You can use the function below to create batches
# but you will have to pay attention at how you store embeddings.
# show the first 20 values of the embedding of the word 'miauwen' and of the pseudoword 'lixthless'
# 2 points for correctly encoding words and pseudowords

def chunks(lst, n):

    """Chunks a list into equal chunks containing n elements. Returns a list of lists."""

    chunked = []
    for i in range(0, len(lst), n):
        chunked.append(lst[i:i + n])
    return chunked


In [414]:
embedding_layer = model.get_input_embeddings()

In [415]:
import torch
## encode the words and pseudowords using RobBERT v2.
speed_words_encoded = [tokenizer.encode(word) for word in speed_words]
speed_words_embedded = [embedding_layer(torch.tensor(encoding)).mean(dim=0).detach() for encoding in speed_words_encoded]
df_pseudowords_encoded = [tokenizer.encode(word) for word in df_pseudowords]
df_pseudowords_embedded = [embedding_layer(torch.tensor(encoding)).mean(dim=0).detach() for encoding in df_pseudowords_encoded]

df_words_robbert = pd.DataFrame(data={"word":speed_words,"encoded":speed_words_encoded,"embedded":speed_words_embedded})
df_pseudowords_robbert = pd.DataFrame(data={"word":df_pseudowords,"encoded":df_pseudowords_encoded,"embedded":df_pseudowords_embedded})

df_words_robbert.head(5)


,word,encoded,embedded
0,mama,"[0, 182, 4828, 2]","[tensor(0.0450), tensor(-0.0272), tensor(-0.00..."
1,ja,"[0, 1732, 2]","[tensor(0.0360), tensor(0.0092), tensor(0.0186..."
2,papa,"[0, 15638, 115, 2]","[tensor(-0.0531), tensor(-0.0346), tensor(0.05..."
3,nee,"[0, 10405, 2]","[tensor(0.0728), tensor(-0.0266), tensor(0.068..."
4,kaka,"[0, 264, 14155, 2]","[tensor(0.0027), tensor(-0.0084), tensor(0.039..."


In [417]:
## Showing the first 20 values of the embedding of the word 'miauwen' and of the pseudoword 'lixthless'
print("miauwen:", end=" ")
print([emb for emb in df_words_robbert[df_words_robbert.word=="miauwen"].embedded][0][:20])

print("lixthless:", end=" ")
print([emb for emb in df_pseudowords_robbert[df_pseudowords_robbert.word=="lixthless"].embedded][0][:20])

miauwen: tensor([ 0.0119, -0.0402, -0.0168, -0.0037,  0.0011,  0.0389,  0.0520, -0.0395,
         0.0395,  0.0163,  0.0040, -0.0408,  0.0064,  0.0133, -0.0244,  0.0518,
         0.0041, -0.0140,  0.0164,  0.0002])
lixthless: tensor([-0.0094,  0.0285, -0.0351,  0.0121, -0.0027,  0.0226,  0.0316, -0.0216,
        -0.0110,  0.0286, -0.0056, -0.0005, -0.0565,  0.0561, -0.0280, -0.0020,
         0.0358, -0.0205,  0.0261, -0.0103])


In [418]:
df_words_robbert.embedded

0        [tensor(0.0450), tensor(-0.0272), tensor(-0.00...
1        [tensor(0.0360), tensor(0.0092), tensor(0.0186...
2        [tensor(-0.0531), tensor(-0.0346), tensor(0.05...
3        [tensor(0.0728), tensor(-0.0266), tensor(0.068...
4        [tensor(0.0027), tensor(-0.0084), tensor(0.039...
                               ...                        
23981    [tensor(0.0260), tensor(-0.0018), tensor(-0.02...
23982    [tensor(-0.0575), tensor(0.0041), tensor(0.003...
23983    [tensor(0.0229), tensor(-0.0503), tensor(0.013...
23984    [tensor(-0.0368), tensor(0.0018), tensor(0.004...
23985    [tensor(-0.0287), tensor(-0.0427), tensor(0.01...
Name: embedded, Length: 23986, dtype: object

In [419]:
# train regression model on word valence estimates from Speed and Brysbaert (2024)
# 1 point for correctly training the regression model
lm_embeddings = LinearRegression()
lm_embeddings.fit([elem.numpy() for elem in df_words_robbert.embedded],speed_valence)

LinearRegression()

In [ ]:
# apply the trained model to predict the valence of pseudowords from Gatti et al (2024).
# Then apply the same model back onto the training set to see how well it predicts the valence of words in Speed and Brysbaert (2024).
# 1 point for correctly applied model

lm_gatti_embed_valence_pred = lm_embeddings.predict([elem.numpy() for elem in df_pseudowords_robbert.embedded])
print(lm_gatti_embed_valence_pred)
lm_speed_embed_valence_pred = lm_embeddings.predict([elem.numpy() for elem in df_words_robbert.embedded])
print(lm_speed_embed_valence_pred)

[0.5838645  0.5437405  0.5146114  ... 0.6065527  0.5653015  0.51663107]
[0.6430716  0.63815165 0.6167155  ... 0.55935717 0.58212894 0.6033806 ]


In [421]:
# compute the Spearman correlation coefficients between true valence and predicted valence for
# - words from Speed and Brysbaert (2024)
# - pseudowords from Gatti and colleagues (2024)
# show the correlation coefficient
# 1 point for the correct Spearman correlation coefficients (rounded to the third decimal place)

In [422]:
print(spearmanr(lm_speed_embed_valence_pred, speed_valence))
print(spearmanr(lm_gatti_embed_valence_pred, true_pseudoword_valence))

SignificanceResult(statistic=np.float64(0.5188058398213267), pvalue=np.float64(0.0))
SignificanceResult(statistic=np.float64(0.18096257852364414), pvalue=np.float64(1.679176726088268e-12))


**Task 4** (*16 points available, 4 for each question*)

Answer the following questions.

**4a.** Describe the performance of each featurization, comparing
- the performance of a same model between the training and test set
- the performance of different models on the training set
- the performance of different models on the test set

(*4 points available, max 150 words*)

>On training set word unigram model had performed worse. Fasttext performed best with a strong relationship (rs = 0.724, p = 0.0) with RoBERT model coming in second with a moderate relationship (rs = 0.51, p = 0.0). 

>On test set unigram model had the highest association (rs = 0.330, p = 4.964 e-25), despite poor performance on training set. The worst performing model was fasttext with a very weak correlation (rs = 0.176, p =  6.046e-12).

>For all models, apart from unigram and bigram models, performance on the training set was greater than the test set. For the word unigram model, performance on words was very weak (rs = 0.087, p = 5.823 e-42), while being only weak for pseudowords (rs = 0.330, p = 4.964 e-25). For bigrams, performance was weak for both train (rs = 0.313, 2.429e-39) and test (rs = 0.318, p = 1.094e-36) data.


**4b.** Compare the correlations you found when training uni-gram, bi-gram, and fastText models on Dutch words and the correlations of similar models trained on English data as reported by Gatti and colleagues; summarize the most important similarities and differences.

(*4 points available, max 150 words*)

>For our models as well as the models presented by Gatti et al. the order of best to worst performance on words was the same. The fasttext model performed the best with a strong correlation (pur model: rs = 0.724, Gatti et al.: rs = 0.79). Similarly, for word bi-grams both models had a weak statistic (rs = 0.313 for our model and rs = 0.33 for Gatti et al.). The worst performing model on both Dutch and English words was the word unigram model. However for the Dutch words the correlation was worse (rs = 0.087) compared to the English words (rs = 0.11).

**4c.** Do you think the performance of the fastText featurization would change if you were to use different n-grams? Would you make them smaller or larger? Justify your answer.

(*4 points available, max 150 words*)

>Larger n-grams might be more specific and therefore allow the model to learn more information from the data provided. Both when trained on Dutch and English, the linear models performed better when presented with bi-grams than with uni-grams, therefore larger n-grams, when passed into the fasttext model which also encodes dimensional information, might lead to better performance on predicting word valence. However, this might also lead to overfitting as the model would be given many more data points to parse through. This, more sparse, dataset would therefore most likely perform much worse when predicting on never before seen data, such as pseudowords. This is already the case for our pseudoword prediction, where larger n-grams lead to worse spearman r values on the test set. 

**4d.** Do you think that training the same models on uni-grams, bi-grams, fastText and transformer-based embeddings but using valence ratings for Finnish (a language which uses the same alphabet as English but is not a IndoEuropean language) words would yield a similar pattern of results? Justify your answer.

(*4 points available, max 150 words*)

>While it is possible that using the same alphabet would yield similar results, it is much more likely that this would not be the case. As Finnish and Dutch do not share common roots, their word structure and word origins might be very different. As such, a rule about the valence of Dutch words might not apply to Finnish. This is also why pseudowords must be language specific, because in some languages certain phonological and grammatical rules forbid the formation of sounds that other languages would permit. As such valence of Finnish words and therefore also Finnish pseudowords might be dependent on completely different aspects of language than the Dutch or English words and pseudowords are.

**Task 5** (*3 points available*)

Compute the average Levenshtein Distance (aLD) between each pseudoword and the 20 words at the smallest edit distance from it. Consider the set of words you used to filter out pseudowords that happen to be valid Dutch words (the file is available in this OSF repository: https://osf.io/9zymw/) to retrieve the 20 words at the smallest edit distance.

In [423]:
# compute the average Levenshtein distance from each pseudoword to the words used to filter out pseudowords.
# Show the aLD estimate for the pseudowords 'nedukes', 'pewbin', and 'vibcines'
# 3 points for correctly computing aLD for pseudowords

In [424]:
## This code is taken from the first notebook discussed in class. Permission to used materials discussed in class has been given
## as long as proper reference is given.

## This code calculates the edit distance
def edit_distance(source, target, costs={'del': 1, 'ins': 1, 'sub':1}):

    r, c = len(source)+1, len(target)+1
    d = np.zeros((r, c), dtype=int)

    for i in range(1, r):
        for j in range(1, c):
            d[i][0] = i
            d[0][j] = j

    for col in range(1, c):
        for row in range(1, r):
            cost = 0 if source[row-1] == target[col-1] else costs['sub']
            d[row][col] = min(d[row-1][col] + costs['del'],
                              d[row][col-1] + costs['ins'],
                              d[row-1][col-1] + cost)

    return int(d[-1][-1])

In [ ]:
## this code calculates the average edit distance of the 20 most similar words for 1 pseudoword
def average_edit_distance(pseudoword, dutch_words):
    edit_distance_list = []
    for word in dutch_words:
        edit_distance_value = edit_distance(pseudoword, word)
        edit_distance_list.append(edit_distance_value)

        if len(edit_distance_list) > 20:
            edit_distance_list.remove(max(edit_distance_list))

    return sum(edit_distance_list)/20


## input a file, returns how many lines there are in the file, so we know which pseudoword to continue with
def retrieve_index(filename):
    open_file = open(filename, "r")
    amount_of_lines = sum(1 for line in open_file)
    open_file.close()
    return amount_of_lines


## input a filename, pseudowords and edit distances will be stored (comma separated) on a new line
def store_in_file(filename):
    amount_of_lines = retrieve_index(filename)
    batch_size = 5 ##amount of elements generated before the file gets closed and re-opened (to save results)
    count = 0

    if amount_of_lines == len(df_pseudowords):
        return "all elements are already written to the file"

    open_file = open(filename, "a") 
    for pseudoword in df_pseudowords[amount_of_lines:]:

        ## closing and re-opening the file to save
        if count >= batch_size:
            open_file.close()
            open_file = open(filename, "a") 
            count = 0
        count += 1

        string = pseudoword + "," + str(average_edit_distance(pseudoword, df_Dutch_words)) + "\n"                                    
        open_file.write(string)

    open_file.close()

    return "all elements have been written to the file"

store_in_file("edit_distance.txt")

'all elements are already written to the file'

In [426]:
## retrieve the values for 'nedukes', 'pewbin', and 'vibcines'
open_file = open("edit_distance.txt", "r")

for i in range(retrieve_index("edit_distance.txt")):
    line = open_file.readline()
    
    if "nedukes" in line or "pewbin" in line or "vibcines" in line:
        pseudoword, edit_distance_value = line.split(",")
        print (f"pseudoword: {pseudoword}, average edit distance: {edit_distance_value}")

pseudoword: nedukes, average edit distance: 2.9

pseudoword: pewbin, average edit distance: 2.95

pseudoword: vibcines, average edit distance: 3.55



**Task 6** (*3 points available*)

For each pseudoword, record the number of tokens in which RobBERT v2 encodes it.

In [427]:
# record the number of tokens in which RobBERT divides each pseudoword
# show the number of tokens for the pseudowords 'yuxwas', 'skibfy', and 'errords'
# 3 points for correctly mapping pseudowords to number of tokens

In [428]:
dictionary_tokencount_robbert = {}
for pseudoword in df_pseudowords:
  dictionary_tokencount_robbert[pseudoword] = len(tokenizer.tokenize(pseudoword))

print("yuxwas:", dictionary_tokencount_robbert["yuxwas"])
print("skibfy:", dictionary_tokencount_robbert["skibfy"])
print("errords:", dictionary_tokencount_robbert["errords"])

yuxwas: 3
skibfy: 4
errords: 3


**Task 7** (*5 points available, see breakdown below*)

Compute the residuals of the predicted valence under the four regressors trained and applied in tasks 2 to 4. Then, correlate the residuals from all four models with aLD. Finally, correlate the residuals from the RobBERT v2 model with the number of tokens in which each pseudoword is split. Use the Pearson's correlation coefficient.

In [ ]:
# compute the residuals from all four regression models fitted before
# 1 point available for correctly computing residuals

##model 1 lm_unigrams
model1Res = true_pseudoword_valence - pw_unigram_valence_pred

##model 2 lm_bigrams
model2Res = true_pseudoword_valence - pw_bigram_valence_pred

##model 3, task 2 Dutch_valence_FastText
model3Res = true_pseudoword_valence - pseudo_valence_FastText_pred

##model 4, task 3 lm_embeddings
model4Res = true_pseudoword_valence - lm_gatti_embed_valence_pred

print(model1Res.describe(), end="\n\n")
print(model2Res.describe(), end="\n\n")
print(model3Res.describe(), end="\n\n")
print(model4Res.describe(), end="\n\n")


count    1499.000000
mean        0.096426
std         0.159033
min        -0.352307
25%        -0.031289
50%         0.095270
75%         0.234493
max         0.421184
Name: predicted_valence, dtype: float64

count    1499.000000
mean        0.084813
std         0.153880
min        -0.419234
25%        -0.029816
50%         0.082573
75%         0.207200
max         0.581062
Name: predicted_valence, dtype: float64

count    1499.000000
mean        0.093721
std         0.161196
min        -0.314343
25%        -0.036231
50%         0.092052
75%         0.227212
max         0.565186
Name: predicted_valence, dtype: float64

count    1499.000000
mean        0.104484
std         0.161847
min        -0.361514
25%        -0.017087
50%         0.103936
75%         0.238050
max         0.478235
Name: predicted_valence, dtype: float64



In [ ]:
# compute the Pearson's correlation between residuals and average LD for all models,
# as well as the correlation between RobBERT v2 residuals and the number of tokens in which each pseudoword
#    is encoded by the RobBERT v2 model.
# show all correlation coefficients
# 4 points for the correct correlation coefficients

from scipy.stats import pearsonr

dfResLD = pd.read_csv('edit_distance.txt', names = ['word', 'LD'])
dfResLD["num_tokens"] = [len(item) for item in df_pseudowords_robbert.encoded]

pearson1 = pearsonr(model1Res, dfResLD['LD'])
pearson2 = pearsonr(model2Res, dfResLD['LD'])
pearson3 = pearsonr(model3Res, dfResLD['LD'])
pearson4 = pearsonr(model4Res, dfResLD['LD'])

print("Model 1 (LD):", pearson1)
print("Model 2 (LD):", pearson2)
print("Model 3 (LD):", pearson3)
print("Model 4 (LD):", pearson4)

## RobBERT residuals vs num_tokens
print("Model 4 (num_tokens):", pearsonr(model4Res,dfResLD.num_tokens))


Model 1 (LD): PearsonRResult(statistic=np.float64(-0.3725717612529925), pvalue=np.float64(1.455322839360907e-50))
Model 2 (LD): PearsonRResult(statistic=np.float64(-0.4086369605433632), pvalue=np.float64(2.0438462601626322e-61))
Model 3 (LD): PearsonRResult(statistic=np.float64(-0.3775370513454169), pvalue=np.float64(5.601984892218653e-52))
Model 4 (LD): PearsonRResult(statistic=np.float64(-0.3528036327248185), pvalue=np.float64(3.5844831673718285e-45))
Model 4 (num_tokens): PearsonRResult(statistic=np.float64(-0.20122903664689767), pvalue=np.float64(3.689339970000726e-15))


**Task 8** What is the relation between the errors each model made and aLD? what about the number of tokens (limited to the RobBERT v2 model)?

(*4 points available, max 150 words*)

>There is a moderate significant inverse relationship between the residuals of the models and the aLD for all 4 models, strongest of which is Model 2 (bigram) (r=-0.409; p<0.001). This indicates that, if the average edit distance increases (and the pseudoword is thus more unique), errors tend to decrease. This shows that words that look more like those in the training data are more likely to result in larger errors.

>Furthermore, for the RobBERT model, a similar, yet slightly weaker, inverse relationship can be observed between the number of tokens and the residuals (r=-0.201; p<0.001). This shows that, as the number of tokens increases, the models tend to make less errors.

